### PR TITLE
[10.x] Add `Macroable` trait to `Sleep` class

### PR DESCRIPTION
**Pull request by bradietilley**
From https://github.com/laravel/framework/pull/47092

This PR adds the Macroable trait to the Sleep class. Example:

```php
// AppServiceProvider
Sleep::macro('forConfiguredTime', static function () {
    $milliseconds = DB::table('settings')->where('key', 'sleep_after_action')->first()->value ?? 1000; // let's say 3500 ms

    return Sleep::for($milliseconds)->milliseconds();
} );

// Throughout your application:
Sleep::forConfiguredTime(); // e.g. 3500ms

// Or
Sleep::for(1)->second()->orMyCustomMacroAmountOfTime();
```

A helpful subfeature for this facility would be the ability to define a replacement duration. Example:

```php
$sleep = Sleep::for(1)->second();

// Replace the duration
$sleep->duration(800)->milliseconds();

// $sleep = 800ms not 1s
```

### CODE REVIEW

File: src/Illuminate/Support/Sleep.php [hunk 1]
```diff
@@ -116,6 +107,32 @@ public static function sleep($duration)
         return (new static($duration))->seconds();
     }
 
+    /**
+     * Sleep for the given duration. Replaces any previously defined duration.
+     *
+     * @param  \DateInterval|int|float  $duration
+     * @return $this
+     */
+    public function duration($duration)
```
Comment (nunomaduro) [inline src/Illuminate/Support/Sleep.php:116]: Seems that this method can (and should) be protected. Or I am missing something?

**Review by nunomaduro — CHANGES_REQUESTED** [review]

File: src/Illuminate/Support/Sleep.php (hunk 1 above)
Comment (bradietilley) [reply src/Illuminate/Support/Sleep.php:116]: The idea is that in a Macro (or Conditionable -- see other PR) you can replace the duration of the sleep from outside of the `$this` scope, without spawning a brand new Sleep instance (as doing so would destruct the other, right?).

Comment (nunomaduro) [reply src/Illuminate/Support/Sleep.php:116]: The macro can access protected methods - as the given closure is bound to the sleep instance. As such, regarding this PR only, the method should be protected.

**Review by nunomaduro — APPROVED** [review]

**Comment by nunomaduro** [comment]
Ups! You will need to adjust your `testItCanReplacePreviouslyDefinedDurations` test - one idea is create a macro to override the duration

**Review by nunomaduro — APPROVED** [review]
I've fixed the test suite.

**Pull request merged**